File: design_patterns/state/orcamento.py
from abc import ABCMeta, abstractclassmethod


class Estado_de_um_orcamento(object):
    __metaclass__ = ABCMeta

    @abstractclassmethod
    def aplica_desconto_extra(self, orcamento):
        pass

    @abstractclassmethod
    def aprova(self, orcamento):
        pass

    @abstractclassmethod
    def reprova(self, orcamento):
        pass

    @abstractclassmethod
    def finaliza(self, orcamento):
        pass


class Em_aprovacao(Estado_de_um_orcamento):
    def aplica_desconto_extra(self, orcamento):
        orcamento.adiciona_desconto_extra(orcamento.valor * 0.02)

    def aprova(self, orcamento):
        orcamento.estado_atual = Aprovado()

    def reprova(self, orcamento):
        orcamento.estado_atual = Reprovado()

    def finaliza(self, orcamento):
        raise Exception('Orcamento em aprovacao não pode ser finalizado!')


class Aprovado(Estado_de_um_orcamento):
    def aplica_desconto_extra(self, orcamento):
        print("aplicando desconto de aprovacao")
        orcamento.adiciona_desconto_extra(orcamento.valor * 0.05)

    def aprova(self, orcamento):
        raise Exception('Orcamento ja esta aprovado')

    def reprova(self, orcamento):
        raise Exception('Orcamento ja esta aprovado')

    def finaliza(self, orcamento):
        orcamento.estado_atual = Finalizado()


class Reprovado(Estado_de_um_orcamento):
    def aplica_desconto_extra(self, orcamento):
        raise Exception('Orcamentos reprovados não terão descontos extras!')

    def aprova(self, orcamento):
        raise Exception('Orcamento foi Reprovado, nao pode ser Aprovado')

    def reprova(self, orcamento):
        raise Exception('Orcamento ja esta REaprovado')

    def finaliza(self, orcamento):
        orcamento.estado_atual = Finalizado()


class Finalizado(Estado_de_um_orcamento):
    def aplica_desconto_extra(self, orcamento):
        raise Exception('Orcamentos finalizados não terão descontos extras!')

    def aprova(self, orcamento):
        raise Exception('Orcamento foi finalizado, nao pode ser aprovado')

    def reprova(self, orcamento):
        raise Exception('Orcamento foi finalizado, nao pode ser REprovado')

    def finaliza(self, orcamento):
        raise Exception('Orcamento foi finalizado')


class Orcamento():
    def __init__(self):
        self.__items = []
        self.estado_atual = Em_aprovacao()
        self.__desconto_extra = 0
        self.__descontos_aplicados = []

    def aprova(self):
        self.estado_atual.aprova(orcamento)

    def reprova(self):
        print('Orcamento Reprovado')
        self.estado_atual.reprova(orcamento)

    def finaliza(self):
        print('Orcamento Finalizado')
        self.estado_atual.finaliza(orcamento)

    def aplica_desconto_extra(self):
        self.estado_atual.aplica_desconto_extra(self)

    def adiciona_desconto_extra(self, desconto):
        if not (self.estado_atual in self.__descontos_aplicados):
            print("desconto = {}".format(desconto))
            self.__desconto_extra += desconto
            self.__descontos_aplicados.append(self.estado_atual)
            print(self.__descontos_aplicados)
        else:
            print("Ja aplicado descontos extras")

    @property
    def valor(self):
        total = 0.0
        for item in self.__items:
            total += item.valor
        return total - self.__desconto_extra

    def obter_items(self):
        return tuple(self.__items)

    @property
    def total_items(self):
        return len(self.__items)

    def adiciona_item(self, item):
        self.__items.append(item)


class Item(object):

    def __init__(self, nome, valor):
        self.__nome = nome
        self.__valor = valor

    @property
    def valor(self):
        return self.__valor


if __name__ == '__main__':
    orcamento = Orcamento()
    orcamento.adiciona_item(Item('Item1', 100))
    orcamento.adiciona_item(Item('Item1', 50))
    orcamento.adiciona_item(Item('Item1', 400))

    print(orcamento.valor)
    orcamento.aplica_desconto_extra()
    print(orcamento.valor)

    # orcamento.reprova()
    # orcamento.aplica_desconto_extra()

    orcamento.aprova()
    orcamento.aplica_desconto_extra()
    orcamento.aplica_desconto_extra()
    print(orcamento.valor)

File: design_patterns/state/orcamento.rb
class EstadoDeUmOrcamento

  def aplica_desconto_extra(_orcamento)
    raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
  end

  def aprova(_orcamento)
    raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
  end

  def reprova(_orcamento)
    raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
  end

  def finaliza(_orcamento)
    raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
  end
end

class EmAprovacao < EstadoDeUmOrcamento
  def aplica_desconto_extra(orcamento)
    orcamento.adiciona_desconto_extra(orcamento.valor * 0.02)
  end

  def aprova(orcamento)
    orcamento.estado_atual = Aprovado.new
  end

  def reprova(orcamento)
    orcamento.estado_atual = Reprovado.new
  end

  def finaliza(orcamento)
    raise Exception, "Orcamento em aprovacao nao pode ser finalizado"
  end

end

class Aprovado < EstadoDeUmOrcamento
  def aplica_desconto_extra(orcamento)
    puts "Aplicando desconto de aprovacao"
    orcamento.adiciona_desconto_extra(orcamento.valor * 0.05)
  end

  def aprova(orcamento)
    raise Exception, "Orcamento ja esta aprovado"
  end

  def reprova(orcamento)
    raise Exception, "Orcamento ja esta aprovado"
  end

  def finaliza(orcamento)
    orcamento.estado_atual = Finalizado.new
  end
end

class Reprovado < EstadoDeUmOrcamento
  def aplica_desconto_extra(orcamento)
    raise Exception, "Orcamentos reprovados nao pode ter descontos"
  end

  def aprova(orcamento)
    raise Exception, "Orcamento ja esta Reprovado"
  end

  def reprova(orcamento)
    raise Exception, "Orcamento ja  foi Reprovado"
  end

  def finaliza(orcamento)
    orcamento.estado_atual = Finalizado.new
  end
end

class Finalizado < EstadoDeUmOrcamento
  def aplica_desconto_extra(orcamento)
    raise Exception, "Orcamentos finaçozados nao pode ter descontos"
  end

  def aprova(orcamento)
    raise Exception, "Orcamento ja esta Finalizado"
  end

  def reprova(orcamento)
    raise Exception, "Orcamento ja esta Finalizado"
  end

  def finaliza(orcamento)
    raise Exception, "Orcamento ja esta Finalizadp"
  end
end

class Orcamento
  attr_reader :valor
  attr_accessor :estado_atual

  def initialize
    @items = []
    @estado_atual = EmAprovacao.new
    @desconto_extra = 0
    @descontos_aplicados = []
  end

  def aprova
    self.estado_atual.aprova(self)
  end

  def reprova
    puts "Orcamento Reprovado"
    self.estado_atual.reprova(self)
  end

  def finaliza
    self.estado_atual.finaliza(self)
  end

  def aplica_desconto_extra
    self.estado_atual.aplica_desconto_extra(self)
  end

  def adiciona_desconto_extra(desconto)
    if @descontos_aplicados.include? self.estado_atual
      puts "Ja aplicado desconto extra"
      return
    end
    puts "adicionando desconto #{desconto}"
    @desconto_extra += desconto
    @descontos_aplicados << self.estado_atual
  end

  def valor
    # return @valor unless @valor.nil?

    @valor = 0
    @items.each { |i| @valor += i.valor }
    puts @desconto_extra
    puts "-------------"
    @valor - @desconto_extra
  end

  def qtd_items
    @items.length
  end

  def adiciona_item(item)
    @items << item
  end
end

class Item
  attr_reader :valor

  def initialize(nome, valor)
    @nome  = nome
    @valor = valor
  end
end

orcamento = Orcamento.new
orcamento.adiciona_item(Item.new('Item1',100))
orcamento.adiciona_item(Item.new('Item2',50))
orcamento.adiciona_item(Item.new('Item3',400))

puts orcamento.valor
orcamento.aplica_desconto_extra
puts orcamento.valor

orcamento.aprova
orcamento.aplica_desconto_extra
orcamento.aplica_desconto_extra
puts orcamento.valor
